### PR TITLE
bug fix: singal sent to non-existent process causes exception

### PR DESCRIPTION
**Pull request by assafgi**

### CODE REVIEW

File: agent/talker.py [hunk 1]
```diff
@@ -430,7 +430,16 @@ def sync_progress(self, pipeline, now):
...
+
+        if sys.version_info.major == 3:
+            exception_type = ProcessLookupError
+        else:
+            exception_type = OSError
+
+        try:
+            os.killpg(os.getpgid(self.popen.pid), sig)
+        except exception_type:
+            return
```
Comment (DoronCohen) [inline agent/talker.py]: lets validate `exception.errno == errno.ESRCH` using `import errno`

File: tests/uts/test_agent.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+def kill_process_before_signaling(self, job_id, signal):
+    job = self.current_processes.get(job_id)
+    os.killpg(os.getpgid(job.popen.pid), signal)
+    while job.popen.poll() is None:
+        pass
+    job.send_signal(signal)
+
+
+JOBS_DIR = '/tmp/talker/jobs'
+JOBS_SEEN = '/tmp/talker/jobs/eos.json'
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: I would expect that only patching `JOBS_DIR` will be sufficient. let's fix that in the code

File: tests/uts/test_agent.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+@patch('agent.talker.JOBS_DIR', JOBS_DIR)
+@patch('agent.talker.JOBS_SEEN', JOBS_SEEN)
+class TestClient(unittest.TestCase):
+
+    def setUp(self):
+        super(TestClient, self).setUp()
+        if not os.path.isdir(JOBS_DIR):
+            os.makedirs(JOBS_DIR)
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: sparse the code

File: tests/uts/test_agent.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        self.agent = TalkerAgent()
+        self.agent.host_id = get_uuid()
+        self.agent.redis = FakeStrictRedis()
+        self.agent.pending_exception = None
+        self.agent_thread = Thread(target=self.agent.start)
+        self.agent_thread.start()
+
+    def tearDown(self):
+        self.agent.kill()
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: you should also join the thread

File: tests/uts/test_agent.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        job_data = {'id': job_id, 'cmd': cmd, 'timeout': 3600.0}
+        self.agent.redis.rpush(jobs_key, json.dumps(job_data))
+        return job_id
+
+    @patch('agent.talker.TalkerAgent.signal_job', kill_process_before_signaling)
+    def test_signal_existed_process(self):
+        job_id = self.run_cmd_on_agent(['bash', '-ce', 'sleep 5'])
+        job_id = self.run_cmd_on_agent('signal', job_id)
+        self.agent.redis.blpop('result-{}-retcode'.format(job_id))
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: dont you want to validate the return code?

File: tests/uts/test_agent.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    job.send_signal(signal)
+
+
+JOBS_DIR = '/tmp/talker/jobs'
+JOBS_SEEN = '/tmp/talker/jobs/eos.json'
+
+
+@patch('agent.talker.JOBS_DIR', JOBS_DIR)
+@patch('agent.talker.JOBS_SEEN', JOBS_SEEN)
+class TestClient(unittest.TestCase):
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: maybe it's worth it to shorten somehow the `RedisFetcher` loop interval to get shorter tests. maybe not in this iteration

File: tests/uts/test_agent.py (hunk 4 above)
Comment (assafgi) [reply tests/uts/test_agent.py]: It is a bit problematic. `join` blocks, if we use `join` it means that we wait for the `agent_thread` to finish, this happens only in the `tearDown`. Do you want to put it after `self.agent.kill()`?

File: tests/uts/test_agent.py (hunk 6 above)
Comment (assafgi) [reply tests/uts/test_agent.py]: The test is quiet fast. The loop `fetch_new_jobs` doesn't have any sleep time, what do you mean by making it shorter?  how do you want to do it?

File: talker_agent/talker.py [hunk 7]
```diff
@@ -430,7 +430,11 @@ def sync_progress(self, pipeline, now):
 
     def send_signal(self, sig):
         self.logger.debug("Sending signal %s", sig)
-        os.killpg(os.getpgid(self.popen.pid), sig)
+        try:
+            os.killpg(os.getpgid(self.popen.pid), sig)
+        except (OSError, ProcessLookupError) as e:
```
Comment (DoronCohen) [inline talker_agent/talker.py]: is `ProcessLookupError` defined in Python 2? why not check the Python version?

File: talker_agent/talker.py [hunk 8]
```diff
@@ -524,6 +529,10 @@ def check_if_seen(self, job_id):
             return
         return first_seen
 
+    def kill(self):
+        self.stop_fetching.set()
```
Comment (DoronCohen) [inline talker_agent/talker.py]: great

File: tests/uts/test_agent.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    def tearDown(self):
+        self.agent.kill()
+        self.agent_thread.join()
+
+    def run_agent(self):
+        try:
+            self.agent.start()
+        except Exception as e:
+            self.exception_class = e.__class__
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: lets be more specific. keep the exception itself and reraise it inside a `assertRaises`

File: tests/uts/test_agent.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+from threading import Thread
+from time import sleep, time
+
+from fakeredis import FakeStrictRedis
+from mock import patch
+
+from talker_agent.talker import TalkerAgent
+
+WAIT_FOR_EXCEPTION_TIMEOUT = 1
+WAIT_FOR_EXCEPTION_ITERATION_SLEEP = 0.0001
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: too short imo

File: tests/uts/test_agent.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        job_id = self.run_cmd_on_agent(['bash', '-ce', 'sleep 5'])
+        job_id = self.run_cmd_on_agent('signal', job_id)
+        _, ret_code = self.agent.redis.blpop('result-{}-retcode'.format(job_id))
+        self.assertEqual(ret_code.decode('utf-8'), '-15')
+
+    @patch('os.killpg', raise_file_not_found)
+    def test_signal_job_unexpected_errno(self):
+        job_id = self.run_cmd_on_agent(['bash', '-ce', 'sleep 5'])
+        job_id = self.run_cmd_on_agent('signal', job_id)
+        self.assert_agent_exception(FileNotFoundError)
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: `FileNotFoundError` != `OSError`

File: tests/uts/test_agent.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+    def run_agent(self):
+        try:
+            self.agent.start()
+        except Exception as e:
+            self.exception = e
+
+    def raise_agent_thread_exception(self):
+        wait_for_exception_timeout = 1
+        wait_for_exception_iteration_sleep = 0.1
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: I thought you wanted to keep them upper case. I agreed btw

File: tests/uts/test_agent.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        _, ret_code = self.agent.redis.blpop('result-{}-retcode'.format(job_id))
+        self.assertEqual(ret_code.decode('utf-8'), '-15')
+
+    @patch('os.killpg', raise_file_not_found)
+    def test_signal_job_unexpected_errno(self):
+        job_id = self.run_cmd_on_agent(['bash', '-ce', 'sleep 5'])
+        self.run_cmd_on_agent('signal', job_id)
+        with self.assertRaises(OSError):
+            self.raise_agent_thread_exception()
+            self.assertEqual(2, self.exception.errno)
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: Im not sure this line will be executed at all

File: tests/uts/test_agent.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+    def tearDown(self):
+        self.agent.kill()
+        self.agent_thread.join()
+
+    def run_agent(self):
+        try:
+            self.agent.start()
+        except Exception as e:
+            self.exception = e
```
Comment (DoronCohen) [inline tests/uts/test_agent.py]: `self.agent_exception` seems more fitting

**Pull request merged**